natss/pkg/dispatcher/dispatcher.go

```diff
@@ -138,35 +127,32 @@ func (s *SubscriptionsSupervisor) signalReconnect() {
 	}
 }
 
-func createReceiverFunc(s *SubscriptionsSupervisor, logger *zap.SugaredLogger) eventingchannels.ReceiverFunc {
-	return func(ctx context.Context, channel eventingchannels.ChannelReference, event cloudevents.Event) error {
-		logger.Infof("Received event from %q channel", channel.String())
-		// publish to Natss
-		ch := getSubject(channel)
-		message, err := event.MarshalJSON()
-		if err != nil {
-			logger.Errorf("Error during marshaling of the message: %v", err)
-			return err
-		}
+func messageReceiverFunc(s *SubscriptionsSupervisor) eventingchannels.UnbufferedMessageReceiverFunc {
+	return func(ctx context.Context, channel eventingchannels.ChannelReference, message binding.Message, transformers []binding.TransformerFactory, header http.Header) error {
+		s.logger.Info("Received event", zap.String("channel", channel.String()))
+
 		s.natssConnMux.Lock()
 		currentNatssConn := s.natssConn
 		s.natssConnMux.Unlock()
 		if currentNatssConn == nil {
-			logger.Errorf("No Connection to NATSS")
-			return fmt.Errorf("No Connection to NATSS")
+			s.logger.Error("no Connection to NATSS")
+			return errors.New("no Connection to NATSS")
+		}
+		sender, err := natsscloudevents.NewSenderFromConn(*currentNatssConn, getSubject(channel))
```

Comment:
seems like you might want to keep that built up sender around for next time?

Reply:
I wrote a benchmark for comparing building a new sender every time and maintaining a `map[string]*stan.Sender` from a `subject` to a `sender` with an associated `sync.Mutex`, these are the results:

```bash
$ go test -cpu=1,2,4,8,16 -bench=.
goos: linux
goarch: amd64
pkg: banchmarksendercloudevents
BenchmarkCreateSender           18308764                66.7 ns/op
BenchmarkCreateSender-2         25325703                39.8 ns/op
BenchmarkCreateSender-4         43595174                26.5 ns/op
BenchmarkCreateSender-8         49811014                24.3 ns/op
BenchmarkCreateSender-16        47619736                26.4 ns/op
BenchmarkDispatcherMap          55042156                20.6 ns/op
BenchmarkDispatcherMap-2        45462733                25.6 ns/op
BenchmarkDispatcherMap-4        25054641                49.1 ns/op
BenchmarkDispatcherMap-8        16760083                81.6 ns/op
BenchmarkDispatcherMap-16       11845035                87.8 ns/op
```
In addition, we have to remove all senders in the map when the connection is not valid anymore. It's not worth it.